velox/exec/ExchangeClient.cpp

```diff
@@ -170,74 +174,42 @@ void ExchangeClient::request(const RequestSpec& requestSpec) {
   }
 }
 
-int32_t ExchangeClient::countPendingSourcesLocked() {
-  int32_t numPending = 0;
-  for (auto& source : sources_) {
-    if (source->isRequestPendingLocked()) {
-      ++numPending;
-    }
-  }
-  return numPending;
-}
-
-int64_t ExchangeClient::getAveragePageSize() {
-  auto averagePageSize =
-      std::min<int64_t>(maxQueuedBytes_, queue_->averageReceivedPageBytes());
-  if (averagePageSize == 0) {
-    averagePageSize = 1 << 20; // 1 MB.
+std::vector<ExchangeClient::RequestSpec>
+ExchangeClient::pickSourcesToRequestLocked() {
+  if (closed_) {
+    return {};
   }
-
-  return averagePageSize;
-}
-
-int32_t ExchangeClient::getNumSourcesToRequestLocked(int64_t averagePageSize) {
-  // Figure out how many more 'averagePageSize' fit into 'maxQueuedBytes_'.
-  // Make sure to leave room for 'numPending' pages.
-  const auto numPending = countPendingSourcesLocked();
-
-  auto numToRequest = std::max<int32_t>(
-      1, (maxQueuedBytes_ - queue_->totalBytes()) / averagePageSize);
-  if (numToRequest <= numPending) {
-    return 0;
+  std::vector<RequestSpec> requestSpecs;
+  while (!emptySources_.empty()) {
+    auto& source = emptySources_.front();
+    bool ok = source->shouldRequestLocked();
+    VELOX_CHECK(ok);
+    requestSpecs.push_back({std::move(source), 0});
+    emptySources_.pop();
   }
-
-  return numToRequest - numPending;
-}
-
-void ExchangeClient::pickSourcesToRequestLocked(
-    RequestSpec& requestSpec,
-    int32_t numToRequest,
-    std::queue<std::shared_ptr<ExchangeSource>>& sources) {
-  while (requestSpec.sources.size() < numToRequest && !sources.empty()) {
-    auto& source = sources.front();
-    if (source->shouldRequestLocked()) {
-      requestSpec.sources.push_back(source);
+  int64_t availableSpace =
+      maxQueuedBytes_ - queue_->totalBytes() - totalPendingBytes_;
+  while (availableSpace > 0 && !producingSources_.empty()) {
+    auto& source = producingSources_.front();
+    int64_t requestBytes = 0;
+    for (auto bytes : source->remainingBytes()) {
+      availableSpace -= bytes;
+      if (availableSpace <= 0) {
+        break;
+      }
+      requestBytes += bytes;
     }
-    sources.pop();
-  }
-}
-
-ExchangeClient::RequestSpec ExchangeClient::pickSourcesToRequestLocked() {
-  if (closed_ || queue_->totalBytes() >= maxQueuedBytes_) {
-    return {};
-  }
-
-  const auto averagePageSize = getAveragePageSize();
-  const auto numToRequest = getNumSourcesToRequestLocked(averagePageSize);
-
-  if (numToRequest == 0) {
-    return {};
+    if (requestBytes == 0) {
+      VELOX_CHECK_LE(availableSpace, 0);
```

Comment:
Would you double check that if this line throws the error is handled property and doesn't cause the query to get stuck? I remember seeing this not quite working.

Reply:
Yes, we just need to make sure that in case of this error the query fails and not gets stuck. Would you force an error here and confirm the result?

Reply:
The query finishes with an error and no stuck: https://fburl.com/presto/2nv1ci95

Reply:
Thank you for verifying.